File: src/types/playlistUploader.ts
export interface TrackMetadata {
  title: string;
  artist: string;
  titlePinyin: string;
  artistPinyin: string;
  searchTerms: string[];
  duration: number;
  thumbnail: string;
}

export interface Playlist {
  name: string;
  tracks: string[];
}

export type MusicIndex = Record<string, TrackMetadata>;

export type SyncStatus = 'idle' | 'syncing' | 'completed' | 'failed';
export type PlaylistDownloadStatus =
  | 'idle'
  | 'fetching_metadata'
  | 'downloading'
  | 'building_index'
  | 'uploading'
  | 'updating_playlist_js'
  | 'restarting_bot'
  | 'completed'
  | 'failed';

export interface PlaylistUploaderProgress {
  stage: string;
  current: number;
  total: number;
  message: string;
}

export interface SyncResult {
  success: boolean;
  indexEntries: number;
  playlistsCount: number;
  playlistNames: string[];
  error?: string;
}

export interface DownloadResult {
  success: boolean;
  downloaded: number;
  cached: number;
  failed: number;
  total: number;
  indexEntries: number;
  playlistTracks: number;
  playlistName?: string;
  downloadedTrackIds: string[];
  error?: string;
}

export interface UploadResult {
  success: boolean;
  uploadedTracks: number;
  skippedTracks: number;
  playlistUploaded: boolean;
  playlistJsUpdated: boolean;
  botRestarted: boolean;
  error?: string;
}

export interface PlaylistUploaderCompleteEvent {
  success: boolean;
  action?: 'sync' | 'download' | 'upload';
  error?: string;
}

File: src-tauri/src/models/playlist_uploader.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackMetadata {
    pub title: String,                 
    pub artist: String,                
    pub title_pinyin: String,          
    pub artist_pinyin: String,         
    pub search_terms: Vec<String>,     
    pub duration: u32,                 
    pub thumbnail: String,           
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Playlist {
    pub name: String,
    pub tracks: Vec<String>,          
}

pub type MusicIndex = HashMap<String, TrackMetadata>;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum SyncStatus {
    Idle,
    Syncing,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum PlaylistDownloadStatus {
    Idle,
    FetchingMetadata,
    Downloading,
    BuildingIndex,
    Uploading,
    UpdatingPlaylistJs,
    RestartingBot,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlaylistUploaderProgress {
    pub stage: String,
    pub current: u32,
    pub total: u32,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DownloadResult {
    pub success: bool,
    pub downloaded: u32,
    pub cached: u32,
    pub failed: u32,
    pub total: u32,
    pub index_entries: u32,
    pub playlist_tracks: u32,
    pub playlist_name: Option<String>,
    pub new_track_ids: Vec<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncResult {
    pub success: bool,
    pub index_entries: u32,
    pub playlists_count: u32,
    pub playlist_names: Vec<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadResult {
    pub success: bool,
    pub uploaded_tracks: u32,
    pub skipped_tracks: u32,
    pub playlist_uploaded: bool,
    pub playlist_js_updated: bool,
    pub bot_restarted: bool,
    pub error: Option<String>,
}
